opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -170,16 +182,25 @@ where
             ));
         }
 
-        let trackers = match self.trackers.write() {
-            Ok(mut trackers) => {
+        let mut trackers = if let Ok(mut trackers_guard) = self.trackers.write() {
```

Comment:
I think we could make a few improvements here:
1. Take the write lock on `trackers_collect` before taking the write lock on `trackers`. That way we can reduce the amount of work done while holding the write lock on `trackers`.
2. I'm not sure how helpful or actionable the warnings on poisoned lock are. We could omit that and make the whole thing less verbose.
3. Use more intuitive variable names. For example, using `tracker_collect` instead of `trackers` to convey that collect is actually running on `tracker_collect`.

Here is an attempt to address the above-mentioned points:

```rust
if let Ok(mut trackers_collect) = self.trackers_for_collect.write() {
    if let Ok(mut trackers_current) = self.trackers.write() {
        swap(
            trackers_collect.deref_mut(),
            trackers_current.deref_mut(),
        );
        self.count.store(0, Ordering::SeqCst);
    }

    let mut seen = HashSet::new();
    for (attrs, tracker) in trackers_collect.drain() {
        if seen.insert(Arc::as_ptr(&tracker)) {
            dest.push(map_fn(attrs, tracker.clone_and_reset(&self.config)));
        }
    }
}
```
```

Reply:
> Take the write lock on trackers_collect before taking the write lock on trackers. That way we can reduce the amount of work done while holding the write lock on trackers.

:100: thanks!

Reply:
> I'm not sure how helpful or actionable the warnings on poisoned lock are. We could omit that and make the whole thing less verbose.

It is not user-actionable, but without the internal log, we'll be blind if the issue actually hits. We need to revisit all internal logs before stable, and this will be revisited along with many similar areas.
Similar to https://github.com/open-telemetry/opentelemetry-rust/pull/2331#discussion_r1855268026

Reply:
made the changes suggested, and keeping internal logs for now. Resolving.

Reply:
> It is not user-actionable, but without the internal log, we'll be blind if the issue actually hits. We need to revisit all internal logs before stable, and this will be revisited along with many similar areas.

After thinking some more, since we know that the operations we perform under the write lock will not panic, there should be no scope of lock poisoning. So, it's okay to remove them entirely.

Reply:
Good point. We can use `.expect()` or `unwrap` in those places where we can sure of no poisoning.